src/mbgl/renderer/render_layer.cpp

```diff
@@ -12,6 +12,56 @@ namespace mbgl {
 
 using namespace style;
 
+LayerRenderItem::LayerRenderItem(RenderTiles renderTiles_, LayerRenderer renderer_, LayerUploader uploader_, Immutable<style::LayerProperties> evaluatedProperties_, RenderPass renderPass_)
+    : renderTiles(std::move(renderTiles_))
+    , renderer(std::move(renderer_))
+    , uploader(std::move(uploader_))
+    , evaluatedProperties(std::move(evaluatedProperties_))
+    , renderPass(renderPass_) {}
+
+LayerRenderItem::~LayerRenderItem() = default;
+
+LayerRenderItem RenderLayer::createRenderItem() {
+    return {
+        renderTiles,
+        createRenderer(),
```

Comment:
I don't see the benefit of storing function pointer to render() (as in most of the cases it is directly returned by LayerRenderer createRenderer()) - render() is already stored in vtable.
Apart from negligible additional storage requirement (to store std::function) it increases the complexity (subjective).

For those createRenderer() that evaluate layer properties before returning std::function, does this introduce functionality change, e.g. `unevaluated.get<FillPattern>()` called only once instead of getting it called in every render()? If so, given that are render tests are single shot, this would require additional testing for all of the render layers. 

```
LayerRenderer RenderFillLayer::createRenderer() {
    bool drawPattern = !unevaluated.get<FillPattern>().isUndefined();
    bool hasOutlineColor = !unevaluated.get<FillOutlineColor>().isUndefined();

    return [hasOutlineColor, drawPattern](PaintParameters& parameters, const LayerRenderItem& renderItem) {
```

Reply:
@astojilj  the goal of this PR is being able to detach `LayerRenderItem`, which is part of render tree and will be invoked on renderer thread from the `RenderLayer`, which is part of orchestration and will be called on orchestration thread.  So, render item should capture all the data necessary for rendering and these data (after being captured) must never be mutated on orchestration side.
These data include reference to render tiles (render tiles themselves are passed to render tree with source render items), reference to immutable image data to be used for textures creation, reference to immutable layer properties.

Introducing of `std::function` allows us to create `LayerRenderItem` for each layer by value and thus to avoid unneeded heap allocations an every frame.